File: device-service/modules/ktwin/eventstore/eventstore.py
import requests
from ..common import get_event_store_url, KTwinEvent
from cloudevents.http import from_http, to_binary

# Event Store Methods
# Interact with event store

def get_latest_twin_event(twin_interface, twin_instance):
    url = get_event_store_url() + "/api/v1/twin-events/%s/%s/latest" % (twin_interface, twin_instance)
    response = requests.get(url)

    if response.status_code == 404:
        return None

    cloud_event = from_http(response.headers, response.content)
    return KTwinEvent(cloud_event)

def update_twin_event(ktwin_event: KTwinEvent):
    url = get_event_store_url() + "/api/v1/twin-events"
    headers, body = to_binary(ktwin_event.cloud_event)
    response = requests.post(url, data=body, headers=headers)

    if response.status_code != 202:
        raise Exception("Error while updating twin event", response)

    return response

File: golang/pkg/ktwin/keventstore/eventstore.go
package keventstore

import (
	"fmt"
	"net/http"
	"os"

	"github.com/Open-Digital-Twin/ktwin-smart-cities-services/pkg/ktwin"
)

func GetLatestTwinEvent(twinInterface, twinInstance string) (*ktwin.TwinEvent, error) {
	if os.Getenv("ENV") == "local" {
		return nil, nil
	}

	url := fmt.Sprintf("%s/api/v1/twin-events/%s/%s/latest", ktwin.GetEventStoreURL(), twinInterface, twinInstance)

	response, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	if response.StatusCode == http.StatusNotFound {
		return nil, nil
	}

	event := ktwin.NewTwinEvent()
	err = event.HandleResponse(response)

	if err != nil {
		return nil, err
	}

	return event, nil
}

func UpdateTwinEvent(twinEvent *ktwin.TwinEvent) error {
	if os.Getenv("ENV") == "local" {
		return nil
	}

	url := ktwin.GetBrokerURL()
	twinEvent.CloudEvent.SetType(fmt.Sprintf(ktwin.EventStoreGenerated, twinEvent.TwinInterface))
	return ktwin.PostCloudEvent(twinEvent.CloudEvent, url)
}
